### PR TITLE
SofQW3 segfault

### LINKED ISSUE #22191: SofQW3 segfault
There is an out-of-bounds error in the new `calcTrapezoidYIntersections` code introduced into `FractionalRebinning.cpp` in order to make the `SofQW3` and `Rebin2D` algorithms run faster. This leads to a segmentation fault which crashes Mantid (without an error message). 

### Expected behavior

`SofQW3` runs and gives a rebinned workspace

### Actual behavior

MantidPlot crashes without any error message.

### Steps to reproduce the behavior

```
ws = CreateSampleWorkspace(binWidth = 0.75, XMin = 0, XMax = 300, XUnit = 'DeltaE')
ws = ScaleX(ws, -150, "Add")
LoadInstrument(ws, InstrumentName='MARI', RewriteSpectraMap = True)
ws_sqw = SofQW3(ws, '0,10,10', 'Direct', 150, EAxisBinning='60,0.75,90')
```

### Platforms affected

All

### PR DESCRIPTION
**Pull request by mducle**
Fixes an out-of-bounds index error in `FractionalRebinning` which causes a segmentation fault when `SofQW3` is called with certain parameters.

**To test:**

Run the following script:

```
ws = CreateSampleWorkspace(binWidth = 0.75, XMin = 0, XMax = 300, XUnit = 'DeltaE')
ws = ScaleX(ws, -150, "Add")
LoadInstrument(ws, InstrumentName='MARI', RewriteSpectraMap = True)
ws_sqw = SofQW3(ws, '0,10,10', 'Direct', 150, EAxisBinning='60,0.75,90')
```

And check that no segfault occurs.

Fixes #22191. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataObjects/src/FractionalRebinning.cpp [hunk 1]
```diff
@@ -185,8 +185,8 @@ void calcTrapezoidYIntersections(
     ur = V2D(xAxis[x_end], mTop * xAxis[x_end] + cTop);
   }
 
-  const size_t nx = x_end - x_start;
-  const size_t ny = y_end - y_start;
+  const size_t nx = x_end - x_start + 1;
+  const size_t ny = y_end - y_start + 1;
   const double ll_x(ll.X()), ll_y(ll.Y()), ul_y(ul.Y());
   const double lr_x(lr.X()), lr_y(lr.Y()), ur_y(ur.Y());
   // Check if there is only a output single bin and inputQ is fully enclosed
```
Comment (martyngigg) [inline Framework/DataObjects/src/FractionalRebinning.cpp]: Is it correct that the next statement would now only fire if `xend==x_start` ans `y_end==y_start`?

Comment (mducle) [reply Framework/DataObjects/src/FractionalRebinning.cpp]: Doh! I think I completely missed the point. I just looked at the loop in the general code (`calcGeneralIntersection`) and thought that it meant that when `[xy]_start==[xy]_end` meant there is only one bin width, and that I had misunderstood what `[xy]_end` meant. But actually the code before was right. 

What `[xy]_start == [xy]_end` means is that the *bin width* in that direction is zero, and this was what was causing the segmentation fault.

So, actually we can fix the segfault without changing so much of the code. I'll push another commit to do this...

**Review by martyngigg — COMMENTED** [review]
@mducle Are there any extra tests that are worth adding here?

**Review by martyngigg — APPROVED** [review]
The extra protection avoids the crash when encountering a  bin with zero width. Good to go and is safe to include in the patch.

**Pull request merged**